lib/_stream_writable.js

```diff
@@ -255,11 +273,28 @@ Writable.prototype.setDefaultEncoding = function setDefaultEncoding(encoding) {
 };
 
 function decodeChunk(state, chunk, encoding) {
-  if (!state.objectMode &&
-      state.decodeStrings !== false &&
-      typeof chunk === 'string') {
-    chunk = Buffer.from(chunk, encoding);
+  if (state.objectMode)
+    return chunk;
+
+  if (typeof chunk === 'string') {
+    if (state.stringDecoder !== null &&
+        encoding === state.stringDecoder.encoding &&
+        state.stringDecoder.lastNeed === 0)
+      return chunk; // No re-encoding encessary.
+
+    if (state.decodeStrings !== false || state.stringDecoder !== null)
+      chunk = Buffer.from(chunk, encoding);
   }
+
+  if (state.stringDecoder) {
+    // chunk is always a Buffer now.
+    if (state.flushing) {
+      chunk = state.stringDecoder.end(chunk);
+    } else {
+      chunk = state.stringDecoder.write(chunk);
+    }
+  }
+
```

Comment:
One performance concern here is that `decodeChunk` is now no longer inlineable due to the size (> 600 characters). If you add a variable (e.g. `sd`) that points to `state.stringDecoder` (e.g. after the `if (state.objectMode)` check) and use that everywhere, it should cut down the source length enough.